RIOT-OS/RIOT
# cpu/esp32: implement periph/gpio_ll and periph/gpio_ll_irq

### Contribution description

This PR provides the implementation of the Next Generation GPIO Low-Level API as defined in PR #16787 for ESP32. Reworked a bit for `periph/gpio` implementation that use ESP-IDF gpio HAL/LL now.

The PR includes
- the implementation of `periph/gpio_ll`
- the implementation of `periph/gpio_ll_irq`
- ~some cleanups to fix vera++ warnings (no functional changes)~ no longer needed
- ~a small bug fix in handling level interrupts that was found when testing~ no longer needed

<details><summary>The output from test application:</summary>

```
main(): This is RIOT! (Version: 2022.10-devel-48-g739a8-cpu/esp32/periph_gpio_ll)
Test / Hardware Deatils:
========================
Cabling:
  P0.18 (PA18) -- P1.0 (PB0)
  P0.19 (PA19) -- P1.1 (PB1)
Number of pull resistor values supported: 1
Number of drive strengths supported: 4
Number of slew rates supported: 1
Valid GPIO ports:
- PORT 0 (PORT A)
- PORT 1 (PORT B)

Testing gpio_port_pack_addr()
=============================

All OK

Testing gpip_ng_init()
======================

Testing is_gpio_port_num_valid() is true for PORT_OUT and PORT_IN:

Testing input configurations for PIN_IN_0:
Support for input with pull up: yes
state: in, pull: up, schmitt trigger: off, value: on
Support for input with pull down: yes
state: in, pull: down, schmitt trigger: off, value: off
Support for input with pull to bus level: no
Support for floating input (no pull resistors): yes
state: in, pull: none, schmitt trigger: off, value: off

Testing output configurations for PIN_OUT_0:
Support for output (push-pull) with initial value of LOW: yes
state: out-pp, drive: weakest, value: off
Output is indeed LOW: yes
state: out-pp, drive: weakest, value: on
Output can be pushed HIGH: yes
Support for output (push-pull) with initial value of HIGH: yes
state: out-pp, drive: weakest, value: on
Output is indeed HIGH: yes
Support for output (open drain with pull up) with initial value of LOW: yes
state: out-od, drive: weakest, pull: up, schmitt trigger: off, value: off
Output is indeed LOW: yes
Support for output (open drain with pull up) with initial value of HIGH: yes
state: out-od, drive: weakest, pull: up, schmitt trigger: off, value: on
Output is indeed HIGH: yes
Support for output (open drain) with initial value of LOW: yes
state: out-od, drive: weakest, pull: none, schmitt trigger: off, value: off
Output is indeed LOW: yes
Support for output (open drain) with initial value of HIGH: yes
state: out-od, drive: weakest, pull: none, schmitt trigger: off, value: on
state: in, pull: down, schmitt trigger: off, value: off
Output can indeed be pulled LOW: yes
state: in, pull: up, schmitt trigger: off, value: on
Output can indeed be pulled HIGH: yes
Support for output (open source) with initial value of LOW: no
Support for output (open source) with initial value of HIGH: no
Support for output (open source with pull up) with initial value of HIGH: no
Support for output (open source with pull up) with initial value of LOW: no
Support for disconnecting GPIO: yes
Output can indeed be pulled LOW: yes
Output can indeed be pulled HIGH: yes

Testing Reading/Writing GPIO Ports
==================================

testing initial value of 0 after init
...OK
testing setting both outputs_optional simultaneously
...OK
testing clearing both outputs_optional simultaneously
...OK
testing toggling first output (0 --> 1)
...OK
testing toggling first output (1 --> 0)
...OK
testing toggling second output (0 --> 1)
...OK
testing toggling second output (1 --> 0)
...OK
testing setting first output and clearing second with write
...OK
testing setting second output and clearing first with write
...OK
All input/output operations worked as expected

Testing External IRQs
=====================

Testing rising edge on PIN_IN_0
... OK
Testing falling edge on PIN_IN_0
... OK
Testing both edges on PIN_IN_0
... OK
Testing masking of IRQs (still both edges on PIN_IN_0)
... OK
Testing level-triggered on HIGH on PIN_IN_0 (when input is LOW when setting up IRQ)
... OK
Testing level-triggered on HIGH on PIN_IN_0 (when input is HIGH when setting up IRQ)
... OK
Testing level-triggered on LOW on PIN_IN_0 (when input is HIGH when setting up IRQ)
... OK
Testing level-triggered on LOW on PIN_IN_0 (when input is LOW when setting up IRQ)
... OK


TEST SUCCEEDED
```

</details>

### Testing procedure

1. Use a `esp32-wroom-32` boaerd and  connect GPIO18 to GPIO32 and GPIO19 to GPIO33.
2. Compile and flash with following command:
    ```
    PORT_IN=1 PIN_IN_0=0 PIN_IN_1=1 PORT_OUT=0 PIN_OUT_0=18 PIN_OUT_1=19 \
    make BOARD=esp32-wroom-32 -C tests/periph_gpio_ll flash term
    ```
    The test should succeed and should give the expected output.

### Issues/PRs references

~Depends on PR #16787~
